### ISSUE TITLE
[no-unsafe-return] Unable to correctly infer type from a generic-extending interface with wildcard parameter

### ISSUE THREAD

**Issue by LXSMNSYC** [typescript-eslint/typescript-eslint#2432]
- [x ] I have tried restarting my IDE and the issue persists.
- [x ] I have updated to the latest version of the packages.
- [ x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```js
module.exports = {
  "root": true,
  "env": {
    "browser": true,
    "es6": true,
    "node": true
  },
  "extends": [
    'plugin:react/recommended',
    'airbnb',
    'airbnb/hooks',
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:@typescript-eslint/recommended-requiring-type-checking",
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:import/typescript",
  ],
  "globals": {
    "Atomics": "readonly",
    "SharedArrayBuffer": "readonly"
  },
  "parser": '@typescript-eslint/parser',
  "parserOptions": {
    "project": './tsconfig.eslint.json',
    "ecmaVersion": 2018,
    "sourceType": 'module',
  },
  "plugins": [
    '@typescript-eslint',
  ],
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [".ts", ".tsx"]
    },
    "import/resolver": {
      "typescript": {},
    },
  },
  "plugins": [
    "@typescript-eslint"
  ],
  "rules": {
    "import/extensions": [
      "error",
      "ignorePackages",
      {
        "js": "never",
        "jsx": "never",
        "ts": "never",
        "tsx": "never"
      }
    ],
    "import/no-unresolved": "error",
    "@typescript-eslint/no-explicit-any": "off",
    "@typescript-eslint/no-unsafe-assignment": "off",
    "react/jsx-filename-extension": [1, { "extensions": [".jsx", ".tsx"] }],
    "react/jsx-props-no-spreading": "off",
  }
};
```

```TS
interface NestA<T> {
  value: T;
}

interface NestB<T> {
  ref: T;
}

interface Example<T> {
  nest: NestA<NestB<T>>;
}

type Infer<T> = T extends Example<infer U> ? U : T;

function test<T extends Example<any>>(value: T): NestB<Infer<T>> {
  return value.nest.value as NestB<Infer<T>>;
}

function testB<T extends Example<any>>(value: T): Infer<T> {
  return test(value).ref;
}
```

TSConfig
```json
{
  "include": ["packages", "examples"],
  "exclude": ["node_modules"],
  "compilerOptions": {
    "module": "ESNext",
    "lib": ["DOM", "ESNext"],
    "importHelpers": true,
    "declaration": true,
    "sourceMap": true,
    "rootDir": "./packages",
    "strict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "moduleResolution": "node",
    "jsx": "react",
    "esModuleInterop": true,
    "target": "ES2015"
  }
}
```

**Expected Result**

- Should infer the property type of `value.nest.value` with the same type as return type.
- Should infer the generic type for the return type to that of the input type.

**Actual Result**

<img width="699" alt="Screen Shot 2020-08-28 at 1 13 49 PM" src="https://user-images.githubusercontent.com/4783372/91523847-57dc9f80-e930-11ea-9056-65a1aec6ad84.png">

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `^3.7.1` |
| `@typescript-eslint/parser`        | `^3.7.1` |
| `TypeScript`                       | `3.9.2` |
| `ESLint`                           | `7.6.0` |
| `node`                             | `14.4.0` |

**Comment by bradzacher** [comment]
This is a pretty complex bugs. I'm not entirely sure what the problem is. Generic constraints are already a complex business, but generic constraints with inferrable types make this worse.

```ts
interface NestA<T> {
  value: T;
}

interface NestB<T> {
  ref: T;
}

interface Example<T> {
  nest: NestA<NestB<T>>;
}

type Infer<T> = T extends Example<infer U> ? U : T;

declare function test<T extends Example<any>>(value: T): NestB<Infer<T>>;

function testB<T extends Example<any>>(value: T): Infer<T> {
  return test(value).ref;
}
```

To simplify this a bit, I just removed the body of the `test` function. [repl](https://ts-ast-viewer.com/#code/JYOwLgpgTgZghgYwgAgHIQM5gIIB4AqAfMgN4BQyyAbnADYCuEAXMvgNxkC+ZZoksiFOiwAhAsXKUoEGC3ZcefaPCTIAogA84AWwAOtCONIVkITGBbCcuK2KKEO3MmACeulAEkQMaEYC8rMgQGpAgACYY6lp6BrigPlDIAKrEAPzJyHIcZGEQCLRw0sgw9CAIYMAA9iDIkFgEQSEQ4ZGaOvqGcCAuhIQAFDQMzKwAlJbmYl4J4g48JWUV1bUTDcGhEVHtsV09-YOMcmPIU75ExlIQYPRQNXVgA3SMIwB00jCOPK7urObIATtsIA)

I spent ~30 mins looking into this and there's a lot to unpack and understand.
TS ofc can find the type `Infer<T>` of `.ref`, but when we ask it to resolve the generic constraint (`checker.getBaseConstraintOfType`), it returns `any`.
It's not a normal `any` though - it's a special `any`...

A normal `any` will look like this:
```json5
{
  flags: 1,
  id: 1,
  intrinsicName: "any",
  objectFlags: 0,
}
```
The resolved constraint any looks like this:
```json5
{
  flags: 1,
  id: 1,
  intrinsicName: "any",
  objectFlags: 67108864, // ts.TypeFlag.NonPrimitive
}
```

I'm not sure if this is significant. I just don't know enough about the types API to know.

**Comment by LXSMNSYC** [comment]
For an actual implementation, this happened on my repo: https://github.com/LXSMNSYC/scoped-model/blob/master/packages/react-scoped-model/src/hooks/useScopedModelContext.tsx

adding `as ScopedModelModelType<T>` still reports an error for the dependent functions (e.g. `useValue` reports the same thing).

**Comment by tadhgmister** [comment]
is there any particular reason you are using this inferring technique and not just writing your generics like this:

```ts
function test<V>(value: Example<V>): NestB<V> {
    return value.nest.value;
}

function testB<V>(value: Example<V>): V {
    return test(value).ref;
}
```

Or if you actually do need `T` to be generic for what ever reason something like this:
```ts
function test<V, T extends Example<V>>(value: T): NestB<V> {
    return value.nest.value;
}

function testB<V, T extends Example<V>>(value: T): V {
    return test<V,T>(value).ref;
}
```
The warning saying that you are returning an any type is still reasonable because if the type parameter can't be inferred for what ever reason you will end up with returning any which is in fact unsafe.  For instance if the argument was optional this would actually be unsound:

```ts
function test<T extends Example<any>>(value: T): NestB<Infer<T>> {
    return value.nest.value as NestB<Infer<T>>;
}

function testB<T extends Example<any>>(value?: T): Infer<T> | undefined {
    return value === undefined ? undefined : test(value).ref;
}
const a = testB(); // a is any.
```

So I'm not sure in general we could really handle this in full.

**Comment by LXSMNSYC** [comment]
My library has an interface that accepts to generic parameters, and there's a lot of functions that accepts that interface. It would be bad, in my case, to define those parameters instead of just inferring from the interface itself to properly set the type constraint.

**Comment by tadhgmister** [comment]
no, my friend you are unsafely using any here, doesn't matter that it is generic, because it is constrained by `any` it is not sound:

```ts
// notice how T is always assignable to number.
interface Example<T extends number> {
    data: T;
}
type ExtractT<E> = E extends Example<infer T> ? T : E;

// notice how using any lets us say it's a string.
function demo<E extends Example<any>>(val: E): string {
    const data: ExtractT<E> = val.data;
    return data;
}
```

Typescript is quite happy to have the function `demo` always returning a string, and yet with any correct use of `Example` it will return a number. By using `any` in the generic constraint in this way it is fine with `ExtractT<E>` being assignable to `string` because `ExtractT<E>` is `any` and by definition `any` is assignable to anything.

You very much are returning a very unsafe any value my friend.

**Comment by lifeiscontent** [comment]
this also happens with React children.

e.g:

```js
  const options = React.useMemo(
    () => React.Children.map(children, child => child.props.value) ?? [],
    [children]
  );
```

even though the type of children I've specified in my props are correctly typed.

**Comment by bradzacher** [comment]
@lifeiscontent - please open a new issue if you've got something to report.
There's nothing there about unused variables, so that is unrelated to this issue.

**Comment by JoshuaKGoldberg** [comment]
Aside: [this ☝️ bot comment](https://github.com/typescript-eslint/typescript-eslint/issues/2432#issuecomment-1859034768) is _not_ spam. Everything it says is true and accurate. Please add appropriate alt text to any image you post, so that it's usable both for GitHub searches and folks who can't visually perceive it. 🙂 

_(context: it was marked as spam and I don't know how or why)_

**Comment by Vinccool96** [comment]
This looks to no longer be an issue. Can I get a confirmation that it indeed was?
